main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1071 @@
+          val (tpe2, eff2) = visitExp(exp2)
+          c.unifyTypeM(tpe1, tpe2, loc)
+          c.unifyTypeM(tvar, Type.Bool, loc)
+          val resTpe = tvar
+          val resEff = Type.mkUnion(eff1, eff2, loc)
+          (resTpe, resEff)
+
+        case SemanticOp.StringOp.Concat =>
+          val (tpe1, eff1) = visitExp(exp1)
+          val (tpe2, eff2) = visitExp(exp2)
+          c.expectTypeM(expected = Type.Str, actual = tpe1, exp1.loc)
+          c.expectTypeM(expected = Type.Str, actual = tpe2, exp2.loc)
+          c.unifyTypeM(tvar, Type.Str, loc)
+          val resTpe = tvar
+          val resEff = Type.mkUnion(eff1, eff2, loc)
+          (resTpe, resEff)
+
+        case _ => throw InternalCompilerException(s"Unexpected binary operator: '$sop'.", loc)
+      }
+
+      case Expr.IfThenElse(exp1, exp2, exp3, loc) =>
+        val (tpe1, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        val (tpe3, eff3) = visitExp(exp3)
+        c.expectTypeM(expected = Type.Bool, actual = tpe1, exp1.loc)
+        c.unifyTypeM(tpe2, tpe3, loc)
+        val resTpe = tpe3
+        val resEff = Type.mkUnion(eff1, eff2, eff3, loc)
+        (resTpe, resEff)
+
+      case Expr.Stm(exp1, exp2, loc) =>
+        val (tpe1, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        val resTpe = tpe2
+        val resEff = Type.mkUnion(eff1, eff2, loc)
+        (resTpe, resEff)
+
+      case Expr.Discard(exp, loc) =>
+        val (tpe, eff) = visitExp(exp)
+        val resTpe = Type.Unit
+        val resEff = eff
+        (resTpe, resEff)
+
+      case Expr.Let(sym, mod, exp1, exp2, loc) =>
+        val (tpe1, eff1) = visitExp(exp1)
+        c.unifyTypeM(sym.tvar, tpe1, loc)
+        val (tpe2, eff2) = visitExp(exp2)
+        val resTpe = tpe2
+        val resEff = Type.mkUnion(eff1, eff2, loc)
+        (resTpe, resEff)
+
+      case Expr.LetRec(sym, ann, mod, exp1, exp2, loc) =>
+        // Ensure that `exp1` is a lambda.
+        //val a = Type.freshVar(Kind.Star, loc)
+        //val b = Type.freshVar(Kind.Star, loc)
+        //val p = Type.freshVar(Kind.Eff, loc)
+        //val expectedType = Type.mkArrowWithEffect(a, p, b, loc)
+        val (tpe1, eff1) = visitExp(exp1)
+        //c.unifyTypeM(expectedType, tpe1, exp1.loc)
+        c.unifyTypeM(sym.tvar, tpe1, exp1.loc)
+        val (tpe2, eff2) = visitExp(exp2)
+        val resTpe = tpe2
+        val resEff = Type.mkUnion(eff1, eff2, loc)
+        (resTpe, resEff)
+
+      case Expr.Region(tpe, loc) =>
+        val resTpe = tpe
+        val resEff = Type.Pure
+        (resTpe, resEff)
+
+      case Expr.Scope(sym, regionVar, exp, evar, loc) =>
+        c.enterRegionM(regionVar.sym)
+        c.unifyTypeM(sym.tvar, Type.mkRegion(regionVar, loc), loc)
+        val (tpe, eff) = visitExp(exp)
+        c.exitRegionM(evar, eff, loc)
+        val resTpe = tpe
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.Match(exp, rules, loc) =>
```

Comment:
I think you could make a helper function for guards fx. since it justs checks that everything is bool and returns a list of effects. That would help the outer readability

Reply:
similar for visitPatterns, and visitBodies.
```scala
val (tpe, eff) = visitExp(exp)
visitPatterns(rules.map(_.pat), tpe, ...)
val guardEffs = visitGuards(rules.flatMap(_.guard), ...)
val (bodyTpe, bodyEffs) = visitBodies(rules.map(_.exp), ...)
val resTpe = bodyTpe
val resEff = Type.mkUnion(eff :: guardEffs ::: bodyEffs, loc)
```

Reply:
perhaps there can be a helper function for branching expressions. Can be used for if branches, match branches, typematch branches

Reply:
I opted for `visitMatchRule` since it's consistent with what we do in other phases